### ISSUE TITLE
Exported data (`PyAPI_DATA` rather than `PyAPI_FUNC`)

### ISSUE THREAD

**Issue by encukou** [capi-workgroup/problems#80]
Exporting data from shared libraries is problematic, and we should instead export functions that fetch the data.
This issue didn't make it to [PEP 733](https://peps.python.org/pep-0733/) (aside from a passing reference to “non-function symbols”).

I don't recall the reason behind this, and I'd like to have it written down so we avoid cargo-culting. @zooba, could you fill it in here? (Or in the [guidelines draft](https://github.com/capi-workgroup/api-evolution/blob/main/guidelines.rst)?)

**Comment by steve-s** [comment]
In HPy we also chose to avoid exporting data, the reason is that function call is more flexible (can be switched to lazy initialization or on demand computation, for example). On top of that I am not sure how well are exported data and (re-)exporting data supported in non-C languages.

**Comment by zooba** [comment]
Yeah, @steve-s has basically nailed it. My proposed text (but feel free to rewrite/amend):

```
Data exports (``PyAPI_DATA``) should no longer be used to make global variables
available to embedders and extensions. Instead, use a function (``PyAPI_FUNC``)
that returns the value and an error code (which may just be a ``NULL`` return).
Any allocation, copying, or reference counting should be performed within the
function. This protects against future changes, such as to thread safety or
refcounting, allows the addition of deprecation warnings or failure conditions,
and better supports non-C callers.

The C API Working Group may approve exceptions to any of these guidelines.
```

**Comment by bschoenmaeckers** [comment]
I’m working on implementing the limited API for RustPython.  This API has advanced significantly in concealing CPython’s implementation details and is becoming a viable alternative to HPy. However, I’m encountering a problem.  Currently, built-in types are exported using static data exports which are fundamentally incompatible with how RustPython constructs these types.

Therefore, I’m seeking an alternative API design. Would it be beneficial to move away from static type exports and instead expose getter functions for all built-in types?

xref https://github.com/RustPython/RustPython/pull/7562

**Comment by bschoenmaeckers** [comment]
I'm looking for something like [Py_GetConstant](https://docs.python.org/3/c-api/object.html#c.Py_GetConstant), but for build-in types.

**Comment by encukou** [comment]
I'd support that change (but can't currently champion it).

**Comment by zooba** [comment]
`PyObject_GetAttr(PyEval_GetBuiltins(), "name")`?

**Comment by encukou** [comment]
`PyEval_GetBuiltins()` is good for most uses, but it's not a proper C-level solution since users can replace/modify a frame's builtins dict.

**Comment by bschoenmaeckers** [comment]
Please see the discusses at https://discuss.python.org/t/c-api-add-py-getbuildintype-and-py-getbuildintypeborrowed/107010

**Comment by zooba** [comment]
> `PyEval_GetBuiltins()` is good for most uses, but it's not a proper C-level solution since users can replace/modify a frame's builtins dict.

I would argue that the limited API should respect those modifications anyway, since it's supposed to be operating at a Python level. Use the CPython API if you want to do CPython-level stuff.

**Comment by encukou** [comment]
“At a Python level” to a point, but not in the *current execution frame* scope (by default at least).
Dividing by zero should raise `PyExc_ZeroDivisionError`, not `builtins.ZeroDivisionError`.

**Comment by zooba** [comment]
> Dividing by zero should raise `PyExc_ZeroDivisionError`, not `builtins.ZeroDivisionError`.

I don't think you can justify that from a language design perspective - we only do it for performance and safety reasons (and because it's easier for native code to access a native struct than a Python one). Consider:

```python
def div(a, b):
    if b == 0:
        raise ZeroDivisionError()
    return a / b
```

(Also consider the ability to catch the exception, which is only possible if we raise the one that Python code is going to expect.)

Either Python code is the model of the Python language and our C implementation tries to satisfy it, or our C implementation is the model of the Python language and the actual Python language adapts to our reference implementation. Right now, we can justify either, though it makes no sense to not at least try to make the implementation match the language.

As the limited API is supposed/hoped to be the set that any Python implementation can provide, it should behave like the Python language, not the C implementation.

**Comment by encukou** [comment]
`PyEval_GetBuiltins` is a [reflection](https://docs.python.org/3/c-api/reflection.html) function that's inspects the current *Python* execution frame.

In Python, functions don't use their caller's builtins:

```py
def div(a, b):
    if b == 0:
        raise ZeroDivisionError()
    return a / b
    
try:
    eval("div(1, 0)", globals={'__builtins__': {}, 'div': div})
except ZeroDivisionError:
    print("caught!")
```

That won't work if `div` is a C function that uses `PyEval_GetBuiltins` to get the exception -- unless the C code somehow sets up a frame for itself, but we don't even have API for that yet.

**Comment by zooba** [comment]
I think we're probably just looking at an edge case interaction between how we handle frames here, because reassigning `__builtins__` members works differently from overriding it:

```python
def div(a, b):
    if b == 0:
        raise ZeroDivisionError()
    return a / b

orig_ZeroDivisionError = ZeroDivisionError
try:
    __builtins__.ZeroDivisionError = OverflowError
    eval("div(1, 0)", globals={'div': div})
except orig_ZeroDivisionError:
    print("not caught!")
except ZeroDivisionError:
    print("caught!")
```

Accessing `PyExc_ZeroDivisionError` is not going to account for this - it'll still raise the original exception.

So which behaviour is Python? And which is "we didn't manage to get the state machine perfect"? And which behaviour do we want to double-down on by embedding it in more official APIs, thereby making it the de facto specification, even if it wasn't intentional?

These questions are probably above our pay grade though, so we might want to call in a SC consult.

**Comment by encukou** [comment]
Also try with `def div(a, b, *, ZeroDivisionError=ZeroDivisionError):`. You can't really “`raise ZeroDivisionError` at the Python level” without dict-ish *locals*.

IMO, the C API should operate at a “desugared” Python level -- in the same way it skips creating frames, doesn't expose locals in in introspectable dict, and doesn't honor overriding `__build_class__`, it should also avoid Python-level `getattr`/`getitem` here.

`PyEval_GetBuiltins` and `PyImport_ImportModule("builtins")` are available if you want to *emulate Python code*, but if you're coding a `uint64` in C, it should act like a builtin, and raise *the* interpreter-wide built-in `ZeroDivisionError`. (In CPython it's a process-wide `static`, that's an implementation detail this new API can help hide.)
(Part of that is also “refusing to guess” -- there are various ways you can write this at the “Python level”, and the edge cases are clear from the code, once you're looking for them. In C, you can IMO only get that clarity with explicit imports & getattrs.)

We can take it to SC, of course.

**Comment by zooba** [comment]
> Part of that is also “refusing to guess” -- there are various ways you can write this at the “Python level”, and the edge cases are clear from the code, once you're looking for them. In C, you can IMO only get that clarity with explicit imports & getattrs

The critical nuance here though is we're specifically talking about the stable ABI, which is the subset of the ABI that we want to be universal across implementations.

I agree at the CPython level, and anyone coding specifically against the CPython ABI gets to choose which way to approach it, so the question is do we mean to embed that same choice into the ~universal native API.

**Comment by encukou** [comment]
Proposal to replace the majority of these with `Py_GetConstant`, when compiling for the Stable ABI: https://discuss.python.org/t/getting-exceptions-built-in-types-using-py-getconstant/108861
